Issues/reader comments

aerych opened the pull request:
This PR refreshes the reader comment table view cells UI, updating the style to match what is currently in the Calypso reader and replacing the DTCoreText backed WPRichTextView with our new WPRichContentView.

Fixes: #4034
Fixes: #4710
Fixes: #4641
Fixes: #6083
Refs: #5594

To test:
Take the ReaderCommentViewController for a test drive and confirm that all the following work as expected:
- Comments still load :)
- Threaded comments are properly nested up to 4 levels deep.
- You can like a comment.  A liked comment correctly shows its status. 
- You can reply to a comment.
- The comment author can be tapped to view the author's blog when the author's URL is available. 
- Comments still load more as expected.
- The post header and the reply text view are resized when changing orientation, and when multitasking on the iPad. 
- Links in comments are opened correctly.
- Images in comments display correctly and can be previewed. 
- Videos in comments display correctly and can be played. 

Needs review: @kurzee would you be game for the code review?  
cc @jleandroperez 

Once this one lands, there will be a final PR to remove DT and the last of its dependencies. :)

### Before: 

![simulator screen shot nov 30 2016 4 56 49 pm](https://cloud.githubusercontent.com/assets/1435271/20774923/262bd3d6-b71e-11e6-96b1-bdc5d7f551fd.png)


### After: 

![simulator screen shot nov 30 2016 4 55 15 pm](https://cloud.githubusercontent.com/assets/1435271/20774920/2216dc32-b71e-11e6-82ed-c5f6d418f0dd.png)

kurzee commented:
It looks like paragraph spacing is larger than expected, maybe doubled in height:

![simulator screen shot dec 1 2016 2 47 42 pm](https://cloud.githubusercontent.com/assets/1873422/20811908/2ed9aaaa-b7d5-11e6-9b65-f5254a688e1f.png)

kurzee commented:
There is a sizing issue when replying with a comment:

1. Select a post with comments.
2. Add a new comment.
3. The new cell is loaded, but with a broken label height.
4. The cell refreshes itself quickly after, but with a large bottom padding.
5. Reloading the controller and the cells displays the expected height.

![reader-comments-height-issue](https://cloud.githubusercontent.com/assets/1873422/20812036/bcad4d8c-b7d5-11e6-8e6d-6397149a5c2f.gif)

kurzee commented:
When selecting a user who does not have a profile URL set, it can still be selected and the modal presents with an alert error:

![reader-comments-bad-url](https://cloud.githubusercontent.com/assets/1873422/20812249/7614ed84-b7d6-11e6-8b44-dbe357767ea1.gif)

kurzee commented:
>Threaded comments are properly nested up to 4 levels deep.

This may be out of scope, but it gets pretty difficult once you're 4 levels deep on an iPhone. I wonder if we might limit it to three? Or maybe tighten up the nesting widths a bit?

kurzee commented:
Getting a crash when trying to view this post: https://testkurzee.wordpress.com/2016/10/22/test-post-12/

1. Like that post above.
2. Go to Reader, Posts I Like.
3. Refresh and try to select that Post.
4. Crash.

Crash log:

```
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000112ab034b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000011251821e objc_exception_throw + 48
	2   CoreFoundation                      0x0000000112b19265 +[NSException raise:format:] + 197
	3   Foundation                          0x000000010cb8ac42 -[NSRLEArray objectAtIndex:effectiveRange:] + 136
	4   Foundation                          0x000000010cb89d86 -[NSConcreteAttributedString attribute:atIndex:effectiveRange:] + 58
	5   WordPress                           0x000000010a063503 _TFC9WordPress17WPRichContentViews7contentSS + 883
	6   WordPress                           0x000000010a0118d6 _TFC9WordPress26ReaderDetailViewControllerP33_6894C07EBB00C79B78761FE3A823FACE17configureRichTextfT_T_ + 790
	7   WordPress                           0x000000010a00be3e _TFC9WordPress26ReaderDetailViewControllerP33_6894C07EBB00C79B78761FE3A823FACE13configureViewfT_T_ + 222
	8   WordPress                           0x000000010a00788a _TFC9WordPress26ReaderDetailViewController11viewDidLoadfT_T_ + 762
	9   WordPress                           0x000000010a0078c2 _TToFC9WordPress26ReaderDetailViewController11viewDidLoadfT_T_ + 34
	10  UIKit                               0x000000010dfc4c99 -[UIViewController loadViewIfRequired] + 1258
	11  UIKit                               0x000000010dfcb102 -[UIViewController __viewWillAppear:] + 118
	12  UIKit                               0x000000010dff5fbf -[UINavigationController _startCustomTransition:] + 1290
	13  UIKit                               0x000000010e006c34 -[UINavigationController _startDeferredTransitionIfNeeded:] + 697
	14  UIKit                               0x000000010e007dc7 -[UINavigationController __viewWillLayoutSubviews] + 58
	15  UIKit                               0x000000010e1fed6f -[UILayoutContainerView layoutSubviews] + 223
	16  UIKit                               0x000000010dee7f50 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1237
	17  QuartzCore                          0x000000010d5e8cc4 -[CALayer layoutSublayers] + 146
	18  QuartzCore                          0x000000010d5dc788 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	19  QuartzCore                          0x000000010d5dc606 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	20  QuartzCore                          0x000000010d56a680 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 280
	21  QuartzCore                          0x000000010d597767 _ZN2CA11Transaction6commitEv + 475
	22  UIKit                               0x000000010de4e484 _afterCACommitHandler + 346
	23  CoreFoundation                      0x0000000112a54e17 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	24  CoreFoundation                      0x0000000112a54d87 __CFRunLoopDoObservers + 391
	25  CoreFoundation                      0x0000000112a39b9e __CFRunLoopRun + 1198
	26  CoreFoundation                      0x0000000112a39494 CFRunLoopRunSpecific + 420
	27  GraphicsServices                    0x0000000115511a6f GSEventRunModal + 161
	28  UIKit                               0x000000010de23964 UIApplicationMain + 159
	29  WordPress                           0x0000000109d1a52e main + 62
	30  libdyld.dylib                       0x00000001139cb68d start + 1
	31  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderCommentCell.swift:
```diff
@@ -0,0 +1,217 @@
+
+    override var indentationWidth: CGFloat {
+        didSet {
+            leadingContentConstraint.constant = CGFloat(indentationLevel) * indentationWidth
```
Hmm, maybe some way to not duplicate these two identical `leadingContentConstraint.constant` code blocks. Not sure of a clean way though.

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderCommentCell.swift:
```diff
@@ -0,0 +1,217 @@
+        WPStyleGuide.applyReaderCardBylineLabelStyle(timeLabel)
+
+        authorButton.titleLabel?.lineBreakMode = .ByTruncatingTail
+        textView.textContainerInset = UIEdgeInsets(top: 2, left: -4, bottom: -16, right: 0)
```
Maybe some `const` or commenting for why these inset numbers are the value that they are? No biggie though.

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderCommentCell.swift:
```diff
@@ -0,0 +1,217 @@
+        replyButton.setImage(tintedIcon, forState: .Normal)
+        replyButton.setImage(highlightedIcon, forState: .Highlighted)
+
+        let title = NSLocalizedString("Reply", comment: "Title of the reply button.")
```
Would we want the localized description to read with something like `"Reader comments: Title of the reply button."`?

kurzee commented:
I would probably recommend setting opaque background colors on any of the subviews or labels for performance purposes, avoiding blending.

![simulator screen shot dec 1 2016 4 12 34 pm](https://cloud.githubusercontent.com/assets/1873422/20814861/19f4b56a-b7e1-11e6-88f5-e1dee2284b14.png)

kurzee commented:
@aerych I think that's it for this round, let me know when ready for round two!

aerych commented:
> Getting a crash when trying to view this post

Addressed in #6261.  Nice catch.

aerych commented:
Hiya @kurzee!  Thanks for an awesome review.  I *think* everything has been addressed.  A couple of notes tho: 

> It looks like paragraph spacing is larger than expected

I looked at this, and another issue where there were list bullets showing up on blank lines.  The double spacing is due to doubled BR tags in some comments on a blog using the P2/O2 theme.  The bullets are due to BRs being nested inside a UL tag, but not an LI tag.  HTML just shows this as padding but NSAttributedString is converting these to additional list items.  I think we can update content sanitization in a different PR and keep this one more about the comment refresh.  Cool?

> There is a sizing issue when replying with a comment

Should be fixed now.

> When selecting a user who does not have a profile URL set

Fixed!

> it gets pretty difficult once you're 4 levels deep on an iPhone

True, and I wouldn't want to nest any more than that.  I think three is too few.  I don't think changing the nesting width would be a good idea.  I tried reducing it to half as much and it became rather difficult to distinguish depths.  Maybe we give it a try as is an iterate based on user feedback (if any)?

> Getting a crash when trying to view this post

Fixed in #6261

> I would probably recommend setting opaque background colors on any of the subviews.

Done but you won't notice a huge difference with the UITextView or the buttons. 

Ready for another round!

kurzee commented:
Here's some observations while replying with a new comment, some or all might be for another day's work:

- While activating the text field, if I was already scrolled to the bottom, I would expect my scroll position to continue at the bottom once the keyboard slides up.
  - I might also want my scroll position to be exactly where it is when replying to a comment higher up in the list, so I can reference it while typing my reply.
- After sending my comment, I would expect the scroll position to be updated to where my new comment landed.

![reader-comment-reply-scrolling](https://cloud.githubusercontent.com/assets/1873422/21155421/c3417a0e-c137-11e6-9133-e11399fd21f0.gif)

kurzee commented:
We might need some spacing tweaks between the avatar, author and time labels.

iOS:

<img width="210" alt="screen shot 2016-12-13 at 1 32 52 pm" src="https://cloud.githubusercontent.com/assets/1873422/21155755/fd95c948-c138-11e6-9ef8-7e1d339f506d.png">

Web:

![screen shot 2016-12-13 at 1 33 01 pm](https://cloud.githubusercontent.com/assets/1873422/21155721/e0b222d6-c138-11e6-90c8-66a24c3f9aed.png)

- On iOS the avatar is just a bit lower than the author button/label, or more likely they are equally top-aligned.
  - On the web the author label sits just a tad lower than the top edge of the avatar.
- On iOS the time label appears to be vertically centered between the author button and text view.
  - On the web the time is just a bit closer to the author label than the comment text.

kurzee commented:
The reply icon is flipped horizontally on iOS compared to the web. There is also a slight color variation FWIW.

iOS:

<img width="153" alt="screen shot 2016-12-13 at 1 39 12 pm" src="https://cloud.githubusercontent.com/assets/1873422/21155931/a323a2ea-c139-11e6-8708-d41bb1805b4a.png">

Web:

![screen shot 2016-12-13 at 1 39 07 pm](https://cloud.githubusercontent.com/assets/1873422/21155941/a7cda9a8-c139-11e6-9a6d-7463388dae32.png)

kurzee commented:
Two crashes from a post with a ton of comments, DM me for the link:

```
2016-12-13 13:50:29.989 WordPress[65880:1015381] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 0 beyond bounds for empty array'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010ce6b34b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000010c8d321e objc_exception_throw + 48
	2   CoreFoundation                      0x000000010cd9cf1b -[__NSArrayM objectAtIndex:] + 203
	3   UIKit                               0x000000010807ad97 -[_UITableViewUpdateSupport(Private) _setupAnimationsForExistingVisibleCells] + 592
	4   UIKit                               0x00000001080880d8 -[_UITableViewUpdateSupport _setupAnimations] + 92
	5   UIKit                               0x0000000107d1c6aa -[UITableView _updateWithItems:updateSupport:] + 3054
	6   UIKit                               0x0000000107d14acf -[UITableView _endCellAnimationsWithContext:] + 17960
	7   WordPress                           0x0000000103f633a5 -[ReaderCommentsViewController richContentViewDidResizeAttachment:] + 149
	8   WordPress                           0x0000000104179d6a _TPAo.24 + 42
	9   WordPress                           0x0000000104179d99 _TTRXFo_dC9WordPress17WPRichContentView_dT__XFo_oS0__dT__ + 25
	10  WordPress                           0x0000000104179e57 _TTRXFo_oC9WordPress17WPRichContentView_dT__XFo_iS0__iT__ + 23
	11  WordPress                           0x0000000104179f01 _TPA__TTRXFo_oC9WordPress17WPRichContentView_dT__XFo_iS0__iT__ + 81
	12  WordPress                           0x0000000104179c89 _TFC9WordPress17WPRichContentView16tableImageSourcefTGSQCSo18WPTableImageSource_10imageReadyGSQCSo7UIImage_12forIndexPathGSQCSo11NSIndexPath__T_ + 1545
	13  WordPress                           0x0000000104179f74 _TToFC9WordPress17WPRichContentView16tableImageSourcefTGSQCSo18WPTableImageSource_10imageReadyGSQCSo7UIImage_12forIndexPathGSQCSo11NSIndexPath__T_ + 100
	14  WordPress                           0x00000001040a603f __51-[WPTableImageSource processImage:forURL:receiver:]_block_invoke_2 + 543
	15  libdispatch.dylib                   0x000000010db4e0cd _dispatch_client_callout + 8
	16  libdispatch.dylib                   0x000000010db305e5 _dispatch_barrier_sync_f_slow_invoke + 617
	17  libdispatch.dylib                   0x000000010db4e0cd _dispatch_client_callout + 8
	18  libdispatch.dylib                   0x000000010db2e8d6 _dispatch_main_queue_callback_4CF + 406
	19  CoreFoundation                      0x000000010ce2f4f9 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	20  CoreFoundation                      0x000000010cdf4f8d __CFRunLoopRun + 2205
	21  CoreFoundation                      0x000000010cdf4494 CFRunLoopRunSpecific + 420
	22  UIFoundation                        0x000000010f125c1b -[NSHTMLReader _loadUsingWebKit] + 2041
	23  UIFoundation                        0x000000010f126f15 -[NSHTMLReader attributedString] + 22
	24  UIFoundation                        0x000000010f0bd45c _NSReadAttributedStringFromURLOrData + 5779
	25  UIFoundation                        0x000000010f0bbd35 -[NSAttributedString(NSAttributedStringUIFoundationAdditions) initWithData:options:documentAttributes:error:] + 115
	26  WordPress                           0x00000001041416a7 _TTOFCSo25NSMutableAttributedStringcfzT4dataCSo6NSData7optionsGVs10DictionarySSPs9AnyObject__18documentAttributesGVs33AutoreleasingUnsafeMutablePointerGSqCSo12NSDictionary___S_ + 135
	27  WordPress                           0x0000000104139068 _TFCSo25NSMutableAttributedStringCfzT4dataCSo6NSData7optionsGVs10DictionarySSPs9AnyObject__18documentAttributesGVs33AutoreleasingUnsafeMutablePointerGSqCSo12NSDictionary___S_ + 104
	28  WordPress                           0x00000001041386ba _TFC9WordPress19WPRichTextFormatter30attributedStringFromHTMLStringfzTSS25defaultDocumentAttributesGSqGVs10DictionarySSPs9AnyObject____GSqCSo18NSAttributedString_ + 1322
	29  WordPress                           0x00000001042ce5df _TZFE9WordPressCSo18NSAttributedString30attributedStringFromHTMLStringfzTSS25defaultDocumentAttributesGSqGVs10DictionarySSPs9AnyObject____GSqS0__ + 239
	30  WordPress                           0x000000010417490e _TFC9WordPress17WPRichContentViews7contentSS + 654
	31  WordPress                           0x00000001045504ec _TFC9WordPress17ReaderCommentCell13configureTextfT_T_ + 796
	32  WordPress                           0x000000010454f418 _TFC9WordPress17ReaderCommentCell13configureCellfCSo7CommentT_ + 168
	33  WordPress                           0x000000010454f48a _TToFC9WordPress17ReaderCommentCell13configureCellfCSo7CommentT_ + 58
	34  WordPress                           0x0000000103f61e1c -[ReaderCommentsViewController configureCell:atIndexPath:] + 556
	35  WordPress                           0x0000000103f62038 -[ReaderCommentsViewController tableView:cellForRowAtIndexPath:] + 184
	36  WordPress                           0x0000000103f0e6cd -[WPTableViewHandler tableView:cellForRowAtIndexPath:] + 205
	37  UIKit                               0x0000000107d437b5 -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] + 757
	38  UIKit                               0x0000000107d43a13 -[UITableView _createPreparedCellForGlobalRow:willDisplay:] + 74
	39  UIKit                               0x0000000107d1747d -[UITableView _updateVisibleCellsNow:isRecursive:] + 3295
	40  UIKit                               0x0000000107d4cd95 -[UITableView _performWithCachedTraitCollection:] + 110
	41  UIKit                               0x0000000107d335ef -[UITableView layoutSubviews] + 222
	42  UIKit                               0x0000000107c9af50 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1237
	43  QuartzCore                          0x0000000107771cc4 -[CALayer layoutSublayers] + 146
	44  QuartzCore                          0x0000000107765788 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	45  QuartzCore                          0x0000000107765606 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	46  QuartzCore                          0x00000001076f3680 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 280
	47  QuartzCore                          0x0000000107720767 _ZN2CA11Transaction6commitEv + 475
	48  QuartzCore                          0x000000010767bb97 _ZN2CA7Display11DisplayLink14dispatch_itemsEyyy + 611
	49  CoreFoundation                      0x000000010cdfd964 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
	50  CoreFoundation                      0x000000010cdfd5f3 __CFRunLoopDoTimer + 1075
	51  CoreFoundation                      0x000000010cdfd17a __CFRunLoopDoTimers + 250
	52  CoreFoundation                      0x000000010cdf4f01 __CFRunLoopRun + 2065
	53  CoreFoundation                      0x000000010cdf4494 CFRunLoopRunSpecific + 420
	54  GraphicsServices                    0x000000010f69ca6f GSEventRunModal + 161
	55  UIKit                               0x0000000107bd6964 UIApplicationMain + 159
	56  WordPress                           0x0000000103e2b3ee main + 62
	57  libdyld.dylib                       0x000000010db9a68d start + 1
	58  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```
```
2016-12-13 13:50:29.989 WordPress[65880:1015381] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 0 beyond bounds for empty array'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010ce6b34b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000010c8d321e objc_exception_throw + 48
	2   CoreFoundation                      0x000000010cd9cf1b -[__NSArrayM objectAtIndex:] + 203
	3   UIKit                               0x000000010807ad97 -[_UITableViewUpdateSupport(Private) _setupAnimationsForExistingVisibleCells] + 592
	4   UIKit                               0x00000001080880d8 -[_UITableViewUpdateSupport _setupAnimations] + 92
	5   UIKit                               0x0000000107d1c6aa -[UITableView _updateWithItems:updateSupport:] + 3054
	6   UIKit                               0x0000000107d14acf -[UITableView _endCellAnimationsWithContext:] + 17960
	7   WordPress                           0x0000000103f633a5 -[ReaderCommentsViewController richContentViewDidResizeAttachment:] + 149
	8   WordPress                           0x0000000104179d6a _TPAo.24 + 42
	9   WordPress                           0x0000000104179d99 _TTRXFo_dC9WordPress17WPRichContentView_dT__XFo_oS0__dT__ + 25
	10  WordPress                           0x0000000104179e57 _TTRXFo_oC9WordPress17WPRichContentView_dT__XFo_iS0__iT__ + 23
	11  WordPress                           0x0000000104179f01 _TPA__TTRXFo_oC9WordPress17WPRichContentView_dT__XFo_iS0__iT__ + 81
	12  WordPress                           0x0000000104179c89 _TFC9WordPress17WPRichContentView16tableImageSourcefTGSQCSo18WPTableImageSource_10imageReadyGSQCSo7UIImage_12forIndexPathGSQCSo11NSIndexPath__T_ + 1545
	13  WordPress                           0x0000000104179f74 _TToFC9WordPress17WPRichContentView16tableImageSourcefTGSQCSo18WPTableImageSource_10imageReadyGSQCSo7UIImage_12forIndexPathGSQCSo11NSIndexPath__T_ + 100
	14  WordPress                           0x00000001040a603f __51-[WPTableImageSource processImage:forURL:receiver:]_block_invoke_2 + 543
	15  libdispatch.dylib                   0x000000010db4e0cd _dispatch_client_callout + 8
	16  libdispatch.dylib                   0x000000010db305e5 _dispatch_barrier_sync_f_slow_invoke + 617
	17  libdispatch.dylib                   0x000000010db4e0cd _dispatch_client_callout + 8
	18  libdispatch.dylib                   0x000000010db2e8d6 _dispatch_main_queue_callback_4CF + 406
	19  CoreFoundation                      0x000000010ce2f4f9 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	20  CoreFoundation                      0x000000010cdf4f8d __CFRunLoopRun + 2205
	21  CoreFoundation                      0x000000010cdf4494 CFRunLoopRunSpecific + 420
	22  UIFoundation                        0x000000010f125c1b -[NSHTMLReader _loadUsingWebKit] + 2041
	23  UIFoundation                        0x000000010f126f15 -[NSHTMLReader attributedString] + 22
	24  UIFoundation                        0x000000010f0bd45c _NSReadAttributedStringFromURLOrData + 5779
	25  UIFoundation                        0x000000010f0bbd35 -[NSAttributedString(NSAttributedStringUIFoundationAdditions) initWithData:options:documentAttributes:error:] + 115
	26  WordPress                           0x00000001041416a7 _TTOFCSo25NSMutableAttributedStringcfzT4dataCSo6NSData7optionsGVs10DictionarySSPs9AnyObject__18documentAttributesGVs33AutoreleasingUnsafeMutablePointerGSqCSo12NSDictionary___S_ + 135
	27  WordPress                           0x0000000104139068 _TFCSo25NSMutableAttributedStringCfzT4dataCSo6NSData7optionsGVs10DictionarySSPs9AnyObject__18documentAttributesGVs33AutoreleasingUnsafeMutablePointerGSqCSo12NSDictionary___S_ + 104
	28  WordPress                           0x00000001041386ba _TFC9WordPress19WPRichTextFormatter30attributedStringFromHTMLStringfzTSS25defaultDocumentAttributesGSqGVs10DictionarySSPs9AnyObject____GSqCSo18NSAttributedString_ + 1322
	29  WordPress                           0x00000001042ce5df _TZFE9WordPressCSo18NSAttributedString30attributedStringFromHTMLStringfzTSS25defaultDocumentAttributesGSqGVs10DictionarySSPs9AnyObject____GSqS0__ + 239
	30  WordPress                           0x000000010417490e _TFC9WordPress17WPRichContentViews7contentSS + 654
	31  WordPress                           0x00000001045504ec _TFC9WordPress17ReaderCommentCell13configureTextfT_T_ + 796
	32  WordPress                           0x000000010454f418 _TFC9WordPress17ReaderCommentCell13configureCellfCSo7CommentT_ + 168
	33  WordPress                           0x000000010454f48a _TToFC9WordPress17ReaderCommentCell13configureCellfCSo7CommentT_ + 58
	34  WordPress                           0x0000000103f61e1c -[ReaderCommentsViewController configureCell:atIndexPath:] + 556
	35  WordPress                           0x0000000103f62038 -[ReaderCommentsViewController tableView:cellForRowAtIndexPath:] + 184
	36  WordPress                           0x0000000103f0e6cd -[WPTableViewHandler tableView:cellForRowAtIndexPath:] + 205
	37  UIKit                               0x0000000107d437b5 -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] + 757
	38  UIKit                               0x0000000107d43a13 -[UITableView _createPreparedCellForGlobalRow:willDisplay:] + 74
	39  UIKit                               0x0000000107d1747d -[UITableView _updateVisibleCellsNow:isRecursive:] + 3295
	40  UIKit                               0x0000000107d4cd95 -[UITableView _performWithCachedTraitCollection:] + 110
	41  UIKit                               0x0000000107d335ef -[UITableView layoutSubviews] + 222
	42  UIKit                               0x0000000107c9af50 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1237
	43  QuartzCore                          0x0000000107771cc4 -[CALayer layoutSublayers] + 146
	44  QuartzCore                          0x0000000107765788 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	45  QuartzCore                          0x0000000107765606 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	46  QuartzCore                          0x00000001076f3680 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 280
	47  QuartzCore                          0x0000000107720767 _ZN2CA11Transaction6commitEv + 475
	48  QuartzCore                          0x000000010767bb97 _ZN2CA7Display11DisplayLink14dispatch_itemsEyyy + 611
	49  CoreFoundation                      0x000000010cdfd964 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
	50  CoreFoundation                      0x000000010cdfd5f3 __CFRunLoopDoTimer + 1075
	51  CoreFoundation                      0x000000010cdfd17a __CFRunLoopDoTimers + 250
	52  CoreFoundation                      0x000000010cdf4f01 __CFRunLoopRun + 2065
	53  CoreFoundation                      0x000000010cdf4494 CFRunLoopRunSpecific + 420
	54  GraphicsServices                    0x000000010f69ca6f GSEventRunModal + 161
	55  UIKit                               0x0000000107bd6964 UIApplicationMain + 159
	56  WordPress                           0x0000000103e2b3ee main + 62
	57  libdyld.dylib                       0x000000010db9a68d start + 1
	58  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

kurzee commented:
I'm not sure how..... but I was able to produce a bug in which all of the comment cells were white, no content, but still scrollable... I'm unable to reproduce.

It occurred when loading a thread with 20 or so comments and scrolling very quickly down the list. I think somewhere along the way the comments were updating and the cell reload animation broke resulting in all the cells turning white while scrolling. While scrolling back up, after reaching the bottom, there _were_ one or two cells that still had content, surrounded in sad, lonely, white nothingness.

😭 😭 😭 😭 😭 😭

aerych commented:
> Here's some observations while replying with a new comment, some or all might be for another day's work

The behavior is consistent with what we currently have for comments so I'm inclined to let it sit for another day as this patch is more about refreshing the cells.  I'm not sure what expectations the majority of our users would have but I agree there's room for improvement.

> We might need some spacing tweaks between the avatar, author and time labels
> The reply icon is flipped horizontally on iOS compared to the web. There is also a slight color variation FWIW

Keen eyes there yo.  I've adjusted the spacing, flipped the icon 180 (completely missed that this transformation is also being done on the web) and fixed the color. 

> Two crashes from a post with a ton of comments

I think I've found the cause and have a way to sort it out. Will confirm and follow up soon.

aerych commented:
Okay. I think its sorted, as long as we don't see any wonky layout due to deferring updating cell heights til after the tableView stops scrolling.   I'm not seeing any so far but that doesn't mean it isn't there.  We talked about this but just to make it public, as a fallback we can avoid the crashes (and potentially that weird blank cell issue) by just not reusing cells. 

Ready for another round sir!

kurzee commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -236,11 +256,12 @@ extension WPRichContentView: WPTextAttachmentManagerDelegate
-        let height: CGFloat = attachment.height > 0 ? attachment.height : Constants.defaultAttachmentHeight
-        attachment.maxSize = CGSize(width: width, height: height)
+        // Until we have a loaded image use a 1/1 height.  We want a nonzero value
+        // to to avoid an edge case issue where 0 frames are not correctly updated
```
`to to`

kurzee commented:
On iPad, open up comments (at least one with a long thread), background the app:
```
2016-12-15 16:58:17.499 WordPress[23607:1337479] *** Assertion failure in void _prepareForCAFlush(UIApplication *__strong)(), /BuildRoot/Library/Caches/com.apple.xbs/Sources/UIKit_Sim/UIKit-3600.6.21/UIApplication.m:2377
2016-12-15 16:58:17.518 WordPress[23607:1337479] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'unexpected start state'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010a580d4b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x0000000109fe921e objc_exception_throw + 48
	2   CoreFoundation                      0x000000010a584e42 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x00000001045f4599 -[NSAssertionHandler handleFailureInFunction:file:lineNumber:description:] + 166
	4   UIKit                               0x00000001053f3fad _prepareForCAFlush + 499
	5   UIKit                               0x0000000105425601 _beforeCACommitHandler + 19
	6   CoreFoundation                      0x000000010a525267 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	7   CoreFoundation                      0x000000010a5251d7 __CFRunLoopDoObservers + 391
	8   CoreFoundation                      0x000000010a5098a6 CFRunLoopRunSpecific + 454
	9   UIFoundation                        0x000000010c9e5b83 -[NSHTMLReader _loadUsingWebKit] + 2041
	10  UIFoundation                        0x000000010c9e6e7d -[NSHTMLReader attributedString] + 22
	11  UIFoundation                        0x000000010c97cf6b _NSReadAttributedStringFromURLOrData + 5779
	12  UIFoundation                        0x000000010c97b844 -[NSAttributedString(NSAttributedStringUIFoundationAdditions) initWithData:options:documentAttributes:error:] + 115
	13  WordPress                           0x0000000101a601b7 _TTOFCSo25NSMutableAttributedStringcfzT4dataCSo6NSData7optionsGVs10DictionarySSPs9AnyObject__18documentAttributesGVs33AutoreleasingUnsafeMutablePointerGSqCSo12NSDictionary___S_ + 135
	14  WordPress                           0x0000000101a57e78 _TFCSo25NSMutableAttributedStringCfzT4dataCSo6NSData7optionsGVs10DictionarySSPs9AnyObject__18documentAttributesGVs33AutoreleasingUnsafeMutablePointerGSqCSo12NSDictionary___S_ + 104
	15  WordPress                           0x0000000101a574ca _TFC9WordPress19WPRichTextFormatter30attributedStringFromHTMLStringfzTSS25defaultDocumentAttributesGSqGVs10DictionarySSPs9AnyObject____GSqCSo18NSAttributedString_ + 1322
	16  WordPress                           0x0000000101be946f _TZFE9WordPressCSo18NSAttributedString30attributedStringFromHTMLStringfzTSS25defaultDocumentAttributesGSqGVs10DictionarySSPs9AnyObject____GSqS0__ + 239
	17  WordPress                           0x0000000101a92dce _TFC9WordPress17WPRichContentViews7contentSS + 654
	18  WordPress                           0x0000000101e6439c _TFC9WordPress17ReaderCommentCell13configureTextfT_T_ + 796
	19  WordPress                           0x0000000101e632c8 _TFC9WordPress17ReaderCommentCell13configureCellfCSo7CommentT_ + 168
	20  WordPress                           0x0000000101e6333a _TToFC9WordPress17ReaderCommentCell13configureCellfCSo7CommentT_ + 58
	21  WordPress                           0x0000000101881b3c -[ReaderCommentsViewController configureCell:atIndexPath:] + 556
	22  WordPress                           0x0000000101881d58 -[ReaderCommentsViewController tableView:cellForRowAtIndexPath:] + 184
	23  WordPress                           0x000000010182e23d -[WPTableViewHandler tableView:cellForRowAtIndexPath:] + 205
	24  UIKit                               0x0000000105568584 -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] + 757
	25  UIKit                               0x00000001055687e2 -[UITableView _createPreparedCellForGlobalRow:willDisplay:] + 74
	26  UIKit                               0x000000010553c2b0 -[UITableView _updateVisibleCellsNow:isRecursive:] + 3295
	27  UIKit                               0x000000010553cc74 -[UITableView _updateVisibleCellsNow:isRecursive:] + 5795
	28  UIKit                               0x0000000105571b64 -[UITableView _performWithCachedTraitCollection:] + 110
	29  UIKit                               0x00000001055583be -[UITableView layoutSubviews] + 222
	30  UIKit                               0x00000001054bfab8 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1237
	31  QuartzCore                          0x0000000104f94bf8 -[CALayer layoutSublayers] + 146
	32  QuartzCore                          0x0000000104f88440 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	33  QuartzCore                          0x0000000104f882be _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	34  QuartzCore                          0x0000000104f16318 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 280
	35  QuartzCore                          0x0000000104f433ff _ZN2CA11Transaction6commitEv + 475
	36  UIKit                               0x00000001054074ba __83-[UIApplication _createSnapshotContextForScene:withName:performLayoutWithSettings:]_block_invoke.2326 + 33
	37  UIKit                               0x0000000105406716 -[UIApplication _performWithUICACommitStateSnapshotting:] + 785
	38  UIKit                               0x0000000105407220 -[UIApplication _createSnapshotContextForScene:withName:performLayoutWithSettings:] + 1172
	39  UIKit                               0x0000000105408779 __65-[UIApplication _performSnapshotsWithAction:forScene:completion:]_block_invoke_2 + 128
	40  FrontBoardServices                  0x000000010cadd518 -[FBSSceneSnapshotAction _executeNextRequest] + 223
	41  FrontBoardServices                  0x000000010cadd183 -[FBSSceneSnapshotAction executeRequestsWithHandler:completionHandler:expirationHandler:] + 218
	42  UIKit                               0x00000001054086ac __65-[UIApplication _performSnapshotsWithAction:forScene:completion:]_block_invoke + 318
	43  UIKit                               0x0000000105407cb8 -[UIApplication _beginSnapshotSessionForScene:withSnapshotBlock:] + 1180
	44  UIKit                               0x0000000105408530 -[UIApplication _performSnapshotsWithAction:forScene:completion:] + 578
	45  UIKit                               0x000000010540826d -[UIApplication _handleSnapshotAction:forScene:completion:] + 153
	46  UIKit                               0x0000000105403cc7 __102-[UIApplication _handleApplicationDeactivationWithScene:shouldForceExit:transitionContext:completion:]_block_invoke.2106 + 539
	47  UIKit                               0x0000000105403791 __102-[UIApplication _handleApplicationDeactivationWithScene:shouldForceExit:transitionContext:completion:]_block_invoke.2092 + 1393
	48  UIKit                               0x000000010540762f _runAfterCACommitDeferredBlocks + 320
	49  UIKit                               0x00000001053f4273 _cleanUpAfterCAFlushAndRunDeferredBlocks + 566
	50  UIKit                               0x0000000105425746 _afterCACommitHandler + 176
	51  CoreFoundation                      0x000000010a525267 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	52  CoreFoundation                      0x000000010a5251d7 __CFRunLoopDoObservers + 391
	53  CoreFoundation                      0x000000010a509f8e __CFRunLoopRun + 1198
	54  CoreFoundation                      0x000000010a509884 CFRunLoopRunSpecific + 420
	55  GraphicsServices                    0x000000010cf5ea6f GSEventRunModal + 161
	56  UIKit                               0x00000001053fac68 UIApplicationMain + 159
	57  WordPress                           0x000000010174af2e main + 62
	58  libdyld.dylib                       0x000000010b45f68d start + 1
	59  ???                                 0x0000000000000001 0x0 + 1
)
```

kurzee commented:
While restoring state, iPhone and iPad, the state will restore as expected but I do notice this in the debugger log:

```
2016-12-15 17:04:47.610 WordPress[26345:1353534] +[CATransaction synchronize] called within transaction
2016-12-15 17:04:47.649 WordPress[26345:1353534] +[CATransaction synchronize] called within transaction
```

aerych commented:
> While restoring state, iPhone and iPad, the state will restore as expected but I do notice this in the debugger log:

I suspected this was more of a framework issue rather so I built a demo app to test my theory.  The demo app consisted of a UITableViewController with custom cells containing a UITextView that was configured to set its attributedText to an NSAttributedString with an HTML doctype.  I can reproduce the exact console message when restoring state with *just* the above and I can avoid the console message by removing the HTML specification.  I think this would be something we could file a radar for but I'm not sure what other steps we could take. We do need to configure cells during the restoration process and the attributed strings have to html document type.

The other issues have been addressed.  Time for another peek?

kurzee commented:
Got this again 😭😭😭😭😭😭😭😭

![reader-comments-blank-2](https://cloud.githubusercontent.com/assets/1873422/21281045/95f44aca-c3af-11e6-883a-066754256a97.gif)
![reader-comments-blank](https://cloud.githubusercontent.com/assets/1873422/21281046/95f65d7e-c3af-11e6-9580-2dd2d51c272b.gif)

kurzee commented:
Pesky, but can be for another day:

- Background the app with comments open and scroll down at least a bit.
- Foreground, notice the scroll offset shifts upwards.

aerych commented:
The vanishing cells were related to an error logged to the console:
```
CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  *** -[__NSArrayM objectAtIndex:]: index 3 beyond bounds [0 .. 2] with userInfo (null)```
```
Fun times! 
The error doesn't halt execution, but setting an all exceptions breakpoint does allow us to examine the stack.  The stack trace shows what has become a familiar pattern with some of the edgecase errors we've seen: creating an `NSAttributedString` with an HTML document type within a `UITableView beginUpdates/endUpdates` block … or really anything involving a `CATransation`.  This seems like a framework issue. :/ 

We can avoid the glitch by reloading the comment's table view without the use of a beginUpdates/endUpdates block and I've patched things accordingly.  Bonus: it gets rid of an annoying little jump in cell heights that occasionally happens when the controller first appears.

aerych commented:
Ready for (hopefully) one final look @kurzee :)

kurzee commented:
>Bonus: it gets rid of an annoying little jump in cell heights that occasionally happens when the controller first appears.

Yusssss 😃

kurzee commented:
@aerych looks like the like action is now being triggered from the author label/button.

kurzee commented:
@aerych nice fix!! This is really feeling solid now.

Apart from the author/like action, this is good go! :shipit: when ready, and then if you can update https://github.com/wordpress-mobile/WordPress-iOS/pull/6337 I'll give it a spin.

aerych commented:
> looks like the like action is now being triggered from the author label/button

https://i.cloudup.com/bBaFo5hPKi.jpg

aerych commented:
Many many many thanks for the review and excellent testing @kurzee!!! :)

jleandroperez commented:
@aerych absolutely beautiful, congratz on merging this one!!!!
